R/filters.R

```diff
@@ -0,0 +1,57 @@
+#' Main function for data filtering
+#'
+#' @param .data The data to be processed. Must be the list of 2 elements:
+#' data table and metadata table.
+#' @param .method Method of filtering. Implemented methods: by.meta.
+#' @param .query Filtering query. It's a named list of filters that will be applied
+#' to data. Elements of the list are specific for selected method:
+#' - by.meta: names are metadata column headers; each value can be string
+#' (metadata value to select for the column), vector of strings (to select multiple
+#' possible values) or numeric expression in format '>#', '<#', '>=#', '<=#' where
+#' '#' is a number (integer or float, can be negative).
+#'
+#' @examples
+#' data(immdata)
+#'
+#' # Select samples with status "MS"
+#' repFilter(immdata, "by.meta", list(Status = "MS"))
+#'
+#' # Select samples from lanes "A" and "B" with age >= 15
+#' repFilter(immdata, "by.meta", list(Lane = c("A", "B"), Age = ">=15"))
+#' @export repFilter
+repFilter <- function(.data, .method, .query, ...) {
+  switch(tolower(.method),
+    by.meta = filter_by_meta(.data, .query),
+    stop("You entered the wrong method! Please, try again.")
+  )
+}
+
+filter_by_meta <- function(.data, .query) {
+  original_classes <- class(.data)
+  names <- names(.query)
+
+  filtered_meta <- .data$meta
+  for (i in seq_along(names)) {
+    name <- names[i]
+    column_query <- unlist(.query[name])
+    filtered_meta <-
+      if (length(column_query) > 1) {
+        filter(filtered_meta, get(name) %in% column_query)
+      } else if (startsWith(column_query, ">=")) {
+        filter(filtered_meta, get(name) >= as.numeric(substring(column_query, 3)))
+      } else if (startsWith(column_query, "<=")) {
+        filter(filtered_meta, get(name) <= as.numeric(substring(column_query, 3)))
+      } else if (startsWith(column_query, ">")) {
+        filter(filtered_meta, get(name) > as.numeric(substring(column_query, 2)))
+      } else if (startsWith(column_query, "<")) {
+        filter(filtered_meta, get(name) < as.numeric(substring(column_query, 2)))
+      } else {
+        filter(filtered_meta, get(name) == column_query)
+      }
+  }
+  filtered_data <- .data$data[filtered_meta$Sample]
```

Comment:
This code will work if and only if `.data` is a `data.frame`. In immunarch `.data` can be either `data.frame`, `data.table` or instance of database. You should use select function from dplyr package to handle all `.data` types

Reply:
But .data is list with two slots for data list and meta data.frame. This code will work in any case while .data$data is named list - which is standard for immunarch data format.

Reply:
Yes, it doesn't make sense to pass datframes or tables without metadata (`.data$data`) to repFilter; I will add check and error message if `.data` is not a list.

Reply:
But .data$data[] selector may not work in some cases, I will change it to dplyr `select`

Reply:
According to `io.R` sources and `v2_data.Rmd` help, `.data$data` is always a list with sample names as names (and dplyr doesn't support lists), and `.data$meta` is always a dataframe, so `filtered_data <- .data$data[filtered_meta$Sample]` is correct and will always work.